sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -13,24 +13,49 @@ namespace Azure.Core
     /// </summary>
     public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, IComparable<ResourceIdentifier>
     {
+        private enum SpecialType
+        {
+            Subscription,
+            ResourceGroup,
+            Location,
+            Provider
+        }
+
+        private readonly struct ResourceIdentifierParts
+        {
+            public ResourceIdentifierParts(ResourceIdentifier parent, ResourceType resourceType, string resourceName, bool isProviderResource, SpecialType? specialType)
+            {
+                Parent = parent;
+                ResourceType = resourceType;
+                ResourceName = resourceName;
+                IsProviderResource = isProviderResource;
+                SpecialType = specialType;
+            }
+
+            public ResourceIdentifier Parent { get; }
+            public ResourceType ResourceType { get; }
+            public string ResourceName { get; }
+            public bool IsProviderResource { get; }
+            public SpecialType? SpecialType { get; }
+        }
+
         internal const char Separator = '/';
 
-        private static ResourceType TenantResourceType = new ResourceType("Microsoft.Resources/tenants");
-        private static ResourceType SubscriptionResourceType = new ResourceType("Microsoft.Resources/subscriptions");
-        private static ResourceType ResourceGroupResourceType = new ResourceType("Microsoft.Resources/resourceGroups");
-        private static ResourceType ProviderResourceType = new ResourceType("Microsoft.Resources/providers");
         private const string RootStringValue = "/";
         private const string ProvidersKey = "providers";
         private const string SubscriptionsKey = "subscriptions";
         private const string LocationsKey = "locations";
-        private const string ResourceGroupsLowerKey = "resourcegroups";
+        private const string ResourceGroupKey = "resourcegroups";
+
+        private const string SubscriptionStart = "/subscriptions/";
+        private const string ProviderStart = "/providers/";
 
         private string? _stringValue;
```

Comment:
Is this still needed?

Reply:
are you asking about _stringValue?  Yes if you construct with a string we store it so we don't have to spend time reconstructing it from the parts

Reply:
In that case, would be be able to use indexes in the string for the property names? And when the value of the property is asked for, return the string? That way you don't allocate the properties until you need to, but you do the processing to figure out where the boundaries are.

Reply:
Yes this would be possible to do.  This wouldn't get us to zero alloc since we are constructing ResourceType stucts along the way as well.

Would you like me to see how much alloc it saves us vs adding additional time on processing to store all the start / stops for all the pieces?

Reply:
We would need to construct those structs anyway. What you are removing are string allocations (which is why `Span<T>` is so cool).

Reply:
@AlexGhiondea, wrt the structs, does it count as an allocation if it's created on the stack?  My understanding is that creating objects on the stack is much faster than making allocations on the heap, but wondering if my terminology is 100% on this.

Reply:
The numbers shown above are all about the number of bytes allocated on the heap. If a struct is created (on the stack) and never boxed, it won't count towards the heap allocations.